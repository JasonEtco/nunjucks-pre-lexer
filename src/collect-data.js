const parsers = require('./parsers')
const parseChildren = require('./parse-children')

module.exports = async function collectData ({ schema, nodes }, data = {}) {
  for (const node of nodes) {
    switch (node.typename) {
      case 'If': {
        const promises = [
          collectData({ schema, nodes: [node.cond] }, data),
          collectData({ schema, nodes: parseChildren(node.body.children) }, data)
        ]

        if (node.else_) {
          const elseChildren = parseChildren(node.else_.children)
          promises.push(collectData({ schema, nodes: elseChildren }, data))
        }

        await Promise.all(promises)
        continue
      }

      case 'For': {
        // The iterable of the for loop
        await collectData({ schema, nodes: [node.arr] }, data)

        // The body of the for loop
        const bodyChildren = parseChildren(node.body.children)
        await collectData({ schema, nodes: bodyChildren }, data)
        continue
      }

      default: {
        if (!(node.typename in parsers)) continue

        await parsers[node.typename](schema, node, data)
        continue
      }
    }
  }

  return data
}
