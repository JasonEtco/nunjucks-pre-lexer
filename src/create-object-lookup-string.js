function getValues (node, pieces = []) {
  if (node.val && 'value' in node.val) {
    pieces.unshift(node.val.value)
  }

  if ('value' in node) {
    pieces.unshift(node.value)
  }

  // LookupVal and nested object use `node.target`
  if (node.target) {
    return getValues(node.target, pieces)
  }

  // FunCall uses `node.name`
  if (node.name) {
    return getValues(node.name, pieces)
  }

  return pieces
}

module.exports = function createObjectLookupString (node) {
  const values = getValues(node)
  return values.join('.')
}
