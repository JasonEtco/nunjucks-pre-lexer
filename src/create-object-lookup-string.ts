import { Node } from './types'

function getValues (node: Node, pieces: string[] = []): string[] {
  if (node.val && 'value' in node.val) {
    pieces.unshift(node.val.value)
  }

  if ('value' in node) {
    pieces.unshift(node.value)
  }

  // LookupVal and nested object use `node.target`
  if (node.target) {
    return getValues(node.target, pieces)
  }

  // FunCall uses `node.name`
  if (node.name) {
    return getValues(node.name, pieces)
  }

  return pieces
}

export function createObjectLookupString (node: Node) {
  const values = getValues(node)
  return values.join('.')
}
