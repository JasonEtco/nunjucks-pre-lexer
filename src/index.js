const { parser } = require('nunjucks')
const parseChildren = require('./parse-children')
const collectData = require('./collect-data')

module.exports = async function getLibraryData (schema, templateString) {
  const ast = parser.parse(templateString)
  const nodes = parseChildren(ast.children)
  return collectData({ schema, nodes })
}
