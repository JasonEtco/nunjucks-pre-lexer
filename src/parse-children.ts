import { Node } from './types'

/**
 * We care about:
 * LookupVal: `{{ foo.bar }}`, getting a property in an object
 * FunCall: `{{ foo() }}`, a function as a variable
 * Set: `{% set key = foo.bar %}`, setting a variable in a template
 */
const IMPORTANT_TYPES = ['LookupVal', 'FunCall', 'If', 'For', 'Set']

export function parseChildren (children: Node[], nodes: Node[] = []) {
  for (const child of children) {
    if (child.children) {
      parseChildren(child.children, nodes)
      continue
    }

    if (IMPORTANT_TYPES.includes(child.typename)) {
      nodes.push(child)
    }
  }

  return nodes
}
