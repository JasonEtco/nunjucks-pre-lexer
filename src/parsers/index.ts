import { FunCall } from './FunCall'
import { LookupVal } from './LookupVal'
import { Node } from '../types'

interface Parsers {
  FunCall: (schema: object, node: Node, data: any) => any
  LookupVal: (schema: object, node: Node, data: any) => any
}

export type ParserKeys = keyof Parsers

export const parsers: Parsers = {
  FunCall,
  LookupVal
}
