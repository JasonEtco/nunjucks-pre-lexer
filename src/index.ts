import { parser } from 'nunjucks'
import { collectData } from './collect-data'
import { parseChildren } from './parse-children'

export async function lexer (schema: object, templateString: string): Promise<any> {
  const ast = parser.parse(templateString)
  const nodes = parseChildren(ast.children)
  return collectData({ schema, nodes })
}
