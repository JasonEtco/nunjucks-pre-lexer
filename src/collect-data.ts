import { parseChildren } from './parse-children'
import { ParserKeys, parsers } from './parsers'
import { ForType, IfType, Node } from './types'

function isIf (node: Node | IfType): node is IfType {
  return node.typename === 'If'
}

function isFor (node: Node | ForType): node is ForType {
  return node.typename === 'For'
}

interface CollectDataArgs {
  schema: any
  nodes: Array<Node | ForType | IfType>
}

export async function collectData ({ schema, nodes }: CollectDataArgs, data = {}) {
  for (const node of nodes) {
    if (isIf(node)) {
      const promises = [
        collectData({ schema, nodes: [node.cond] }, data),
        collectData({ schema, nodes: parseChildren(node.body.children) }, data)
      ]

      if (node.else_) {
        const elseChildren = parseChildren(node.else_.children)
        promises.push(collectData({ schema, nodes: elseChildren }, data))
      }

      await Promise.all(promises)
      continue
    }

    if (isFor(node)) {
      // The iterable of the for loop
      await collectData({ schema, nodes: [node.arr] }, data)

      // The body of the for loop
      const bodyChildren = parseChildren(node.body.children)
      await collectData({ schema, nodes: bodyChildren }, data)
      continue
    }

    if (!(node.typename in parsers)) continue
    await parsers[node.typename as ParserKeys](schema, node, data)
  }

  return data
}
