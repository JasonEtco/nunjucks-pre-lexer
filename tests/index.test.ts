import { lexer } from '../src'

describe('lexer', () => {
  let schema: any

  beforeEach(() => {
    schema = { lib: { test: true } }
  })

  it('gets the data of an schemaect lookup', async () => {
    const src = '{{ lib.test }}'
    const data = await lexer(schema, src)
    expect(data).toEqual(schema)
  })

  it('gets the return value of a function in the schemaect', async () => {
    const src = '{{ lib.test() }}'
    schema = { lib: { async test () { return 'a value' } } }

    const data = await lexer(schema, src)
    expect(data).toMatchSnapshot()
    expect(data.lib.test()).toBe('a value')
  })

  it('gets the return value of a function with arguments', async () => {
    const src = '{{ lib.test("Example") }}'
    schema = { lib: { async test (str: string) { return str } } }

    const data = await lexer(schema, src)
    expect(data).toMatchSnapshot()
    expect(data.lib.test()).toBe('Example')
  })

  it('gets the return value of a function with arguments that are schemaects', async () => {
    const src = '{{ lib.test({ foo: "Example" }) }}'
    schema = { lib: { async test ({ foo }: { foo: string}) { return foo } } }

    const data = await lexer(schema, src)
    expect(data).toMatchSnapshot()
    expect(data.lib.test()).toBe('Example')
  })

  it('gets the return value of a function with arguments that are arrays', async () => {
    const src = '{{ lib.test([1, 2, 3]) }}'
    schema = { lib: { async test (arr: number[]) { return arr[1] } } }

    const data = await lexer(schema, src)
    expect(data).toMatchSnapshot()
    expect(data.lib.test()).toBe(2)
  })

  it('gets the return value of a function with an argument of nested schemaects', async () => {
    const src = '{{ lib.test({ foo: { bar: { bork: false, baz: true } } }) }}'
    schema = { lib: { async test (o: any) { return o.foo.bar.baz } } }

    const data = await lexer(schema, src)
    expect(data).toMatchSnapshot()
    expect(data.lib.test()).toBe(true)
  })

  it('gets the return value of a non-namespaced function', async () => {
    const src = '{{ test() }}'
    schema = { async test () { return 'hiya' } }

    const data = await lexer(schema, src)
    expect(data).toMatchSnapshot()
    expect(data.test()).toBe('hiya')
  })

  it('gets the value calculated after a function\'s execution in a lookup', async () => {
    const src = '{{ lib.test().example }}'
    schema = {
      lib: {
        async test () {
          return { example: true }
        }
      }
    }

    const data = await lexer(schema, src)
    expect(data.lib.test()).toEqual({ example: true })
  })

  // TODO: Allow for nested function execution
  it.skip('gets the value calculated after a nested function\'s execution in a lookup', async () => {
    const src = '{{ lib.test().example().foo }}'
    schema = {
      lib: {
        async test () {
          return {
            async example () {
              return {
                foo: true
              }
            }
          }
        }
      }
    }

    const data = await lexer(schema, src)
    expect(data.lib.test().example()).toEqual({ foo: true })
  })

  describe('tags', () => {
    describe('set', () => {
      it('gets the value', async () => {
        const src = '{% set example = lib.test %}'
        const data = await lexer(schema, src)
        expect(data.lib.test).toBe(true)
      })
    })

    describe('for', () => {
      it('gets the values in the iterable of a for', async () => {
        const src = '{% for thing in lib.test %}hello{% endfor %}'
        schema = { lib: { test: [true, false] } }

        const data = await lexer(schema, src)
        expect(data).toEqual({ lib: { test: [true, false] } })
      })

      it('gets the values in the body of a for', async () => {
        const src = '{% for thing in things %}{{ lib.test }}{% endfor %}'

        const data = await lexer(schema, src)
        expect(data).toEqual({ lib: { test: true } })
      })
    })

    describe('if', () => {
      it('gets the values in the conditional of an if', async () => {
        const src = '{% if lib.test %}hello{% endif %}'

        const data = await lexer(schema, src)
        expect(data).toEqual({ lib: { test: true } })
      })

      it('gets the values in the body of an if', async () => {
        const src = '{% if true %}{{ lib.test }}{% endif %}'

        const data = await lexer(schema, src)
        expect(data).toEqual({ lib: { test: true } })
      })

      it('gets the values in the body of an else', async () => {
        const src = '{% if true %}true{% else %}{{ lib.test }}{% endif %}'

        const data = await lexer(schema, src)
        expect(data).toEqual({ lib: { test: true } })
      })
    })
  })
})
