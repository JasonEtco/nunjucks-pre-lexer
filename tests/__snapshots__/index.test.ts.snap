// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lexer gets the return value of a function in the schemaect 1`] = `
Object {
  "lib": Object {
    "test": [Function],
  },
}
`;

exports[`lexer gets the return value of a function with an argument of nested schemaects 1`] = `
Object {
  "lib": Object {
    "test": [Function],
  },
}
`;

exports[`lexer gets the return value of a function with arguments 1`] = `
Object {
  "lib": Object {
    "test": [Function],
  },
}
`;

exports[`lexer gets the return value of a function with arguments that are arrays 1`] = `
Object {
  "lib": Object {
    "test": [Function],
  },
}
`;

exports[`lexer gets the return value of a function with arguments that are schemaects 1`] = `
Object {
  "lib": Object {
    "test": [Function],
  },
}
`;

exports[`lexer gets the return value of a non-namespaced function 1`] = `
Object {
  "test": [Function],
}
`;
